bot-components/Git_utils.ml

```diff
@@ -219,3 +222,135 @@ let init_git_bare_repository ~bot_info =
+  | _ ->
+      failwith
+        (f "Could not split repository full name %s into (owner, repo)."
+           github_full_name )
+
+let parse_gitlab_repo_url ~http_repo_url =
+  if
+    not
+      (String_utils.string_match ~regexp:"https?://\\([^/]*\\)/\\(.*/.*\\)"
+         http_repo_url )
+  then
+    Result.Error (f "Could not parse GitLab repository URL %s." http_repo_url)
+  else
+    Result.Ok
+      (Str.matched_group 1 http_repo_url, Str.matched_group 2 http_repo_url)
+
+let parse_gitlab_repo_url_and_print ~http_repo_url =
+  match parse_gitlab_repo_url ~http_repo_url with
+  | Ok (gitlab_domain, gitlab_repo_full_name) ->
+      Stdio.printf "GitLab domain: \"%s\"\n" gitlab_domain ;
+      Stdio.printf "GitLab repository full name: \"%s\"\n" gitlab_repo_full_name
+  | Error msg ->
+      Stdio.print_endline msg
+
+let%expect_test "http_repo_url_parsing_coq" =
+  parse_gitlab_repo_url_and_print ~http_repo_url:"https://gitlab.com/coq/coq" ;
+  [%expect
+    {|
+   GitLab domain: "gitlab.com"
+   GitLab repository full name: "coq/coq" |}]
+
+let%expect_test "http_repo_url_parsing_mathcomp" =
+  parse_gitlab_repo_url_and_print
+    ~http_repo_url:"https://gitlab.inria.fr/math-comp/math-comp" ;
+  [%expect
+    {|
+  GitLab domain: "gitlab.inria.fr"
+  GitLab repository full name: "math-comp/math-comp" |}]
+
+let%expect_test "http_repo_url_parsing_example_from_gitlab_docs" =
+  parse_gitlab_repo_url_and_print
+    ~http_repo_url:"https://gitlab.example.com/gitlab-org/gitlab-test" ;
+  [%expect
+    {|
+  GitLab domain: "gitlab.example.com"
+  GitLab repository full name: "gitlab-org/gitlab-test" |}]
+
+let github_repo_of_gitlab_url ~gitlab_mapping ~http_repo_url =
+  parse_gitlab_repo_url ~http_repo_url
+  |> Result.map ~f:(fun (gitlab_domain, gitlab_repo_full_name) ->
+         github_repo_of_gitlab_project_path ~gitlab_mapping ~gitlab_domain
+           ~gitlab_repo_full_name )
+
+(* TODO: ensure there's no race condition for 2 push with very close timestamps *)
+let mirror_action ~bot_info ?(force = true) ~gitlab_domain ~gh_owner ~gh_repo
+    ~gl_owner ~gl_repo ~base_ref ~head_sha () =
+  (let open Lwt_result.Infix in
+   let local_ref = base_ref ^ "-" ^ head_sha in
+   let gh_ref =
+     {repo_url= f "https://github.com/%s/%s" gh_owner gh_repo; name= base_ref}
+   in
+   (* TODO: generalize to use repository mappings, with enhanced security *)
+   gitlab_repo ~bot_info ~gitlab_domain
+     ~gitlab_full_name:(gl_owner ^ "/" ^ gl_repo)
+   |> Lwt.return
+   >>= fun gl_repo ->
+   let gl_ref = {repo_url= gl_repo; name= base_ref} in
+   git_fetch gh_ref local_ref |> execute_cmd
+   >>= fun () -> git_push ~force ~remote_ref:gl_ref ~local_ref () |> execute_cmd
+  )
+  >>= function
+  | Ok () ->
+      Lwt.return_unit
+  | Error e ->
+      Lwt_io.printlf
+        "Error while mirroring branch/tag %s of repository %s/%s: %s" base_ref
+        gh_owner gh_repo e
+
+let apply_after_label ~bot_info ~owner ~repo ~after ~label ~action ~throttle ()
```

Comment:
This function is related to the feature to close stale pull requests, so it should probably be near it (not sure where it has moved).